hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/connector/WriteObservableP.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.impl.connector;
+
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.jet.core.Inbox;
+import com.hazelcast.jet.core.Outbox;
+import com.hazelcast.jet.core.Processor;
+import com.hazelcast.jet.core.ProcessorSupplier;
+import com.hazelcast.jet.core.Watermark;
+import com.hazelcast.jet.impl.observer.ObservableUtil;
+import com.hazelcast.jet.impl.util.Util;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.ringbuffer.OverflowPolicy;
+import com.hazelcast.ringbuffer.Ringbuffer;
+import com.hazelcast.ringbuffer.impl.RingbufferProxy;
+
+import javax.annotation.Nonnull;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Stream;
+
+import static java.util.stream.Collectors.toList;
+
+public final class WriteObservableP<T> implements Processor {
+
+    private static final int ASYNC_OPS_LIMIT = 1;
+    private static final int MAX_BATCH_SIZE = RingbufferProxy.MAX_BATCH_SIZE;
+
+    private final String ringbufferName;
+    private final List<T> batch = new ArrayList<>(MAX_BATCH_SIZE);
+    private final AtomicInteger pendingWrites = new AtomicInteger(0);
+
+    private Ringbuffer<Object> ringbuffer;
+    private ILogger logger;
+
+
+    private WriteObservableP(String ringbufferName) {
+        this.ringbufferName = ringbufferName;
+    }
+
+    @Override
+    public void init(@Nonnull Outbox outbox, @Nonnull Context context) {
+        HazelcastInstance instance = context.jetInstance().getHazelcastInstance();
+        this.logger = context.logger();
+        this.ringbuffer = instance.getRingbuffer(ObservableUtil.getRingbufferName(ringbufferName));
+    }
+
+    @Override
+    public void process(int ordinal, @Nonnull Inbox inbox) {
+        if (batch.isEmpty()) {
+            inbox.drainTo(batch, MAX_BATCH_SIZE);
+        }
+        tryFlush();
+    }
+
+    @Override
+    public boolean tryProcess() {
+        return tryFlush();
+    }
+
+    private boolean tryFlush() {
```

Comment:
We should also flush in `saveToSnapshot`, otherwise we could miss items in case of a failure.